### PR TITLE
[js/web] JSEP Attention & MultiHeadAttention

### PR DESCRIPTION
**Pull request by dakenf**
### Description
This is a narrow implementation of Attention/MultiHeadAttention as it does not support:
a. inputs 5-7 for MHA
b. packed QKV/KV
c. past/present
d. attention mask

But it works well for StableDiffusion and can be extended later. It reduces VRAM usage as it combines many ops into few
I've updated demo here https://islamov.ai/stable-diffusion-webgpu/ it takes ~13sec for 1 image with 20 steps on RTX3090Ti and about 25s on M1 Pro
VRAM usage is about 8gb if you don't use img2img

Going to focus on SDXL now

### CODE REVIEW

**Comment by satyajandhyala** [comment]
Need to run `npm run build:doc` under onnxruntime\js\web to update documentation?

**Comment by dakenf** [comment]
> Need to run `npm run build:doc` under onnxruntime\js\web to update documentation?

yeah. its been a while since i've submitted a new op and forgot about that. also forgot to run format after moving tests from my working branch

**Comment by fs-eire** [comment]
> > Need to run `npm run build:doc` under onnxruntime\js\web to update documentation?
> 
> yeah. its been a while since i've submitted a new op and forgot about that. also forgot to run format after moving tests from my working branch

please update comments in https://github.com/microsoft/onnxruntime/blob/main/js/web/script/generate-webgpu-operator-md.ts#L10 since it's a partial implementation

**Comment by dakenf** [comment]
> please update comments in https://github.com/microsoft/onnxruntime/blob/main/js/web/script/generate-webgpu-operator-md.ts#L10 since it's a partial implementation

done

**Comment by dakenf** [comment]
fixed ONNX Runtime Web CI Pipeline error with unused variable

**Comment by guschmue** [comment]
Awesome if you can SDXL. wasm64 might be a bit of pain but long term not avoidable and a bunch of people badly want it. I think fp16 together with the onnx external data format will go a long way at one point we need 64bit.

**Comment by dakenf** [comment]
> Awesome if you can SDXL. wasm64 might be a bit of pain but long term not avoidable and a bunch of people badly want it. I think fp16 together with the onnx external data format will go a long way at one point we need 64bit.

I've already loaded it (both fp32 and fp16) but having some issues with either onnx export or pipeline code. Getting NaNs after unet run. Most likely will have time to resolve closer to the next week

**Comment by dakenf** [comment]
Anyway, i can maintain my own package with 64bit build. Since my goal is to have diffusers.js library, not specific implementation details. But right now my dev branch and upstream are very diverged. If you agree to merge things to support 64bit flags then it would make everything so much easier. I can fill some separate PRs, just let me know what you would like to have in upstream and how we can make it compatible

File: js/web/lib/wasm/jsep/webgpu/ops/attention.ts [hunk 1]
```diff
@@ -0,0 +1,634 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {TensorView} from '../../tensor-view';
+import {createAttributeWithCacheKey} from '../attribute-with-cache-key';
+import {ComputeContext, GpuDataType} from '../types';
+
+import {castToF32, fillVector, getMaxComponents, inputVariable, outputVariable, ShaderHelper, sumVector, tensorTypeToWsglStorageType} from './common';
+
+export enum AttentionQkvFormat {
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/attention.ts:10]: please use `const enum` to save the generated file size

File: js/web/lib/wasm/jsep/webgpu/ops/attention.ts [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+        x[offset + i] = ${inputHelper.type.value}(exp(f32input - maxValue) / sum);
+      }
+    }
+  }`;
+
+  context.compute(
+      {
+        name: 'computeAttentionProbsSoftmax',
+        cacheHint: '0',
+        inputTypes: [GpuDataType.default],
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/attention.ts:310]: please merge latest change where ProgramInfoLoader is removed.

Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/ops/attention.ts:310]: need a "getRunData" here. see #17803 for which introduced this change.

**Review by satyajandhyala — APPROVED** [review]

**Pull request merged**